File: java/src/main/java/io/binac/leetcode/LinkedListCycle.java
package io.binac.leetcode;

import io.binac.leetcode.util.ListNode;

/**
 * Given a linked list, determine if it has a cycle in it.
 * <p>
 * <p>Follow up:
 * <p>Can you solve it without using extra space?
 */
public class LinkedListCycle {
    public static class Solution1 {
        public boolean hasCycle(ListNode head) {
            ListNode slow = head, fast = head;
            while (fast != null && fast.next != null) {
                slow = slow.next;
                fast = fast.next.next;
                if (slow == fast)
                    return true;
            }
            return false;
        }
    }
}

File: python3/leetcodepy/linked_list_cycle.py
"""
Given a linked list, determine if it has a cycle in it.

Follow up:
Can you solve it without using extra space?
"""

from .utils import ListNode


class Solution1(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        slow, fast = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                return True
        return False
